--- relay.c	2015-11-13 13:33:26.000000000 +0000
+++ relay-patched.c	2016-07-14 00:24:43.273125665 +0000
@@ -39,6 +39,29 @@
 #include "routerparse.h"
 #include "scheduler.h"
 
+// not tao, but still easy to find w/ the same search
+
+#include "stdlib.h"
+#include "memory.h"
+
+static char* webfp_logfile_path = NULL;
+
+static char*
+get_logfile_path()
+{
+  if (!webfp_logfile_path) {
+    char *base_dir = "/var/log/tor/";
+    char* logfile_path = "tor_cell_seq.log";
+    int x = strlen(base_dir) + strlen(logfile_path) + 1;
+    char* abs_logfile_path = (char*) malloc(x);
+    strcpy(abs_logfile_path, base_dir);
+    strcat(abs_logfile_path, logfile_path);
+    webfp_logfile_path = abs_logfile_path;
+  }
+  return webfp_logfile_path;
+}
+//
+
 static edge_connection_t *relay_lookup_conn(circuit_t *circ, cell_t *cell,
                                             cell_direction_t cell_direction,
                                             crypt_path_t *layer_hint);
@@ -570,6 +593,23 @@
                               size_t payload_len, crypt_path_t *cpath_layer,
                               const char *filename, int lineno)
 {
+  //tao
+
+  FILE* fp;
+  char* abs_logfile_path = get_logfile_path();
+  fp = fopen(abs_logfile_path, "a");
+  if (fp != NULL) {
+    struct timeval tp;
+    gettimeofday(&tp, NULL);
+    fprintf(fp, "%f ", tp.tv_sec + tp.tv_usec/1000000.0);
+    fprintf(fp, "OUTGOING CIRC %u, STREAM %d, ", circ->n_circ_id, stream_id);
+    fprintf(fp, "COMMAND %s(%d), ", relay_command_to_string(relay_command), relay_command);
+    fprintf(fp, "length %zu\n", payload_len);
+    fprintf(fp, "\n");
+    fflush(fp);
+    fclose(fp);
+  }
+
   cell_t cell;
   relay_header_t rh;
   cell_direction_t cell_direction;
@@ -1431,6 +1471,26 @@
   tor_assert(circ);
 
   relay_header_unpack(&rh, cell->payload);
+
+  //tao
+
+  FILE* fp;
+  char* abs_logfile_path = get_logfile_path();
+  fp = fopen(abs_logfile_path, "a");
+
+  if (fp != NULL) {
+    struct timeval tp;
+    gettimeofday(&tp, NULL);
+    fprintf(fp, "%f ", tp.tv_sec + tp.tv_usec/1000000.0);
+    fprintf(fp, "INCOMING CIRC %u, STREAM %d, ", circ->n_circ_id, rh.stream_id);
+    fprintf(fp, "COMMAND %s(%d), ", relay_command_to_string(rh.command), rh.command);
+    fprintf(fp, "length %d\n", rh.length);
+    fprintf(fp, "\n");
+    fflush(fp);
+    fclose(fp);
+  }
+
+
 //  log_fn(LOG_DEBUG,"command %d stream %d", rh.command, rh.stream_id);
   num_seen++;
   log_debug(domain, "Now seen %d relay cells here (command %d, stream %d).",
